### PR TITLE
Add utility to produce .a files

### PR DESCRIPTION
**Pull request by steven-johnson**
Work for #879: add a way to combine .o files into .a files to simplify
downstream consumption of multi-architecture output that can’t easily
be linked directly by LLVM at this time.

### CODE REVIEW

**Comment by steven-johnson** [comment]
(Note that I haven't built/tested on Windows as I don't yet have a build system set up for it; relying on the buildbot to complain for now, which it probably will)

**Comment by steven-johnson** [comment]
Expanded the tests, added deterministic mode (which is default), one more minor fix. Amazingly the Travis checks pass which I assume means we are good on Windows. PTAL.

**Comment by steven-johnson** [comment]
(Oh, also strip paths to just basenames per spec)

**Comment by abadams** [comment]
Travis doesn't actually test on Windows. That lot falls to the build bots.
I should really figure out how to get them to test pull requests.

**Comment by steven-johnson** [comment]
> Travis doesn't actually test on Windows

Doh. Well, I need to get this Windows machine on my desk set up anyway...

File: src/StaticLibrary.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        return { a.st_size, a.st_mtime, a.st_uid, a.st_gid, a.st_mode };
+    }
+
+private:
+    const std::string name;
+    FILE *f;
+};
+
+std::string pad_right(std::string s, size_t max) {
+    internal_assert(s.size() <= max) << s.size() << " "<<s;
```
Comment (abadams) [inline src/StaticLibrary.cpp]: nit: spaces around last <<

File: src/StaticLibrary.cpp [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+#endif
+
+#include "Error.h"
+
+namespace Halide {
+namespace Internal {
+
+namespace {
+
+class File {
```
Comment (abadams) [inline src/StaticLibrary.cpp]: Probably belongs in Util.h or similar so people don't reinvent it next time someone wants to do cross-platform file access.

Comment (steven-johnson) [inline src/StaticLibrary.cpp]: Agreed. Although Zalman pointed out offline that std::ofstream might be a better choice for some of this.

**Comment by steven-johnson** [comment]
Removed File and used fstream instead; moved the static helper wrappers to Util.h as suggested. (If this lands then use of those wrappers will get propagated to existing sites.)

**Pull request merged**